api/server/router/container/container_routes.go

```diff
@@ -562,14 +563,45 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 			config.Healthcheck.StartInterval = 0
 		}
 
-		for _, m := range hostConfig.Mounts {
-			if m.BindOptions != nil {
-				// Ignore ReadOnlyNonRecursive because it was added in API 1.44.
-				m.BindOptions.ReadOnlyNonRecursive = false
-				if m.BindOptions.ReadOnlyForceRecursive {
-					return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
+		// Transform read-only binds into Mounts.
+		// This allows us to default ReadOnlyNonRecursive to true.
+		var newBinds []string
+		parser := mounts.NewParser()
+		for _, b := range hostConfig.Binds {
+			if bind, err := parser.ParseMountRaw(b, hostConfig.VolumeDriver); err == nil {
+				if !bind.RW {
+					mnt := bind.Spec
+					if mnt.BindOptions == nil {
+						mnt.BindOptions = &mount.BindOptions{}
+					}
+					// This is default for the short bind syntax (-v src:dst).
+					mnt.BindOptions.CreateMountpoint = true
+
+					hostConfig.Mounts = append(hostConfig.Mounts, mnt)
+					continue
 				}
 			}
+			newBinds = append(newBinds, b)
+		}
+		hostConfig.Binds = newBinds
+
+		for idx, m := range hostConfig.Mounts {
+			if m.Type == mount.TypeBind && m.ReadOnly {
+				bo := m.BindOptions
+				if bo != nil {
+					if bo.ReadOnlyForceRecursive {
+						return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
+					}
```

Comment:
Perhaps something like this;

```go
for idx, m := range hostConfig.Mounts {
	if m.BindOptions != nil && m.BindOptions.ReadOnlyForceRecursive {
		// NOTE that technically this is a breaking change for older
		// API versions, and we should ignore the new field.
		// However, this option may be incorrectly set by a client with
		// the expectation that the failing to apply recursive read-only
		// is enforced, so we decided to produce an error instead,
		// instead of silently ignoring.
		return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
	}
	if m.Type == mount.TypeBind && m.ReadOnly {
		if m.BindOptions == nil {
			hostConfig.Mounts[idx].BindOptions = &mount.BindOptions{}
		}
		// ReadOnlyNonRecursive was added in API 1.44.
		// Before that all read-only mounts were non-recursive.
		// Keep that behavior for clients on older APIs.
		hostConfig.Mounts[idx].BindOptions.ReadOnlyNonRecursive = true
	}
}
```

Reply:
Hmm right, I guess it doesn't harm to do it for non-readonly case too (even though it doesn't really do anything without ReadOnly).